### PR TITLE
Add built-in deterministic code scorers to `mlflow.genai.scorers`

### LINKED ISSUE #20827: [FR] Built-in deterministic "Tier 1" scorers — rule-based structural & safety checks that run before LLM judges
# [FR] Built-in deterministic "Tier 1" scorers — rule-based structural & safety checks that run before LLM judges

## Willingness to contribute

Yes. I can contribute this feature independently.

## Proposal Summary

MLflow ships 25+ built-in LLM-based scorers (`Correctness`, `Safety`, `RetrievalGroundedness`, etc.) but **zero built-in deterministic code scorers** for structural and safety checks — things like schema validity, invariants, policy gates, and hard constraints.

This proposal introduces a set of **built-in deterministic "Tier 1" scorers** that complement the existing LLM judges. These rule-based checks are:

- **Free** — no LLM API calls
- **Instant** — microseconds, not seconds
- **Reproducible** — identical results every run
- **Gating** — can short-circuit expensive LLM evaluation when outputs are structurally invalid

### Proposed built-in scorers

| Scorer | What it checks | Category |
|---|---|---|
| `ExactMatch` | Output matches expected response | Invariant |
| `JsonValidity` | Output is valid JSON, optionally matching a schema | Schema validity |
| `RegexMatch` | Output matches a regex pattern | Policy gate |
| `ContainsKeywords` | Output contains required keywords/phrases | Policy gate |
| `LengthBound` | Response length within min/max (chars, words, or tokens) | Hard constraint |
| `IsNotEmpty` | Response is non-empty and non-whitespace | Invariant |
| `LatencyThreshold` | Trace latency under a maximum threshold | Hard constraint |
| `NumericBound` | Numeric output within a value range | Hard constraint |

### Two-phase roadmap

**Phase 1 (this proposal)**: Ship the 8 built-in code scorers as first-class peers to LLM judges, usable in `mlflow.genai.evaluate()` today.

**Phase 2 (follow-up)**: Add `gate=True` support so Tier 1 scorers run **before** LLM judges — if a gate fails, skip expensive probabilistic evaluation for that row. This modifies the evaluation harness execution order and saves real cost at scale.

## Motivation

> #### What is the use case for this feature?

Running fast, deterministic structural and safety checks on LLM outputs — either alongside or **before** probabilistic LLM-as-a-Judge evaluation. Examples:

- **Schema validity**: Does the output parse as valid JSON? Does it conform to the expected schema?
- **Policy gates**: Does the response include a required disclaimer? Does it avoid forbidden patterns?
- **Hard constraints**: Is the response under the token limit? Is the confidence score within bounds?
- **Invariants**: Is the response non-empty? Does it match the expected format?

> #### Why is this use case valuable to support for MLflow users in general?

**1. Every production evaluation pipeline needs both tiers.**

In practice, teams always combine deterministic checks with LLM judges. The deterministic checks catch structural failures (malformed JSON, missing fields, empty responses) while LLM judges assess subjective quality (correctness, tone, safety). Today, MLflow provides built-ins only for the second tier.

**2. Tier 1 checks should gate Tier 2 — saving cost.**

If `JsonValidity` fails, there's no point calling `openai:/gpt-4` to judge "correctness" on garbage output. Running deterministic checks first and skipping LLM calls on structural failures saves real money. At 1,000 eval rows × 5 LLM judges × $0.01/call, even a 20% structural failure rate saves $100/run.

**3. The most common checks are undifferentiated boilerplate.**

The `@scorer` decorator documentation shows `exact_match` and `not_empty` as examples users copy-paste. Every team independently reimplements these same patterns, handling edge cases (null outputs, type mismatches, encoding) in different ways. Built-in scorers standardize this.

**4. Code scorers currently produce worse results than LLM judges.**

User-written `@scorer` functions typically return bare `True`/`False` without rationale. LLM judges automatically generate explanations. Built-in code scorers would return well-structured `Feedback` objects with informative rationale (e.g., "Output length 2,847 chars exceeds maximum of 2,000"), giving parity in the evaluation results UI.

> #### Why is this use case valuable to support for your project(s) or organization?

We evaluate LLM outputs in regulated domains where deterministic constraint validation is mandatory — outputs must be valid JSON, contain specific fields, fall within numerical bounds, and meet format requirements before any subjective quality assessment begins. Having these as MLflow built-ins would standardize our pipeline and enable the "gate before LLM judge" pattern we need for cost control.

> #### Why is it currently difficult to achieve this use case?

- **Zero built-in code scorers exist.** `mlflow.genai.scorers` exports 25+ classes — all `BuiltInScorer(Judge)` subclasses requiring an LLM.
- **`BuiltInScorer` is LLM-coupled.** It extends `Judge`, which requires `instructions` (an LLM prompt) and `feedback_value_type`. Code scorers cannot inherit from this hierarchy.
- **No gating mechanism.** All scorers run in parallel in the `ThreadPoolExecutor`. There is no way to say "run these checks first, skip the rest if they fail."
- **Copy-paste problem.** Every team reimplements `exact_match`, `is_valid_json`, `is_not_empty` from the `@scorer` docstring examples.

### Concrete scenarios where this adds immense merit

**Scenario 1: RAG pipeline with structured output**
An RAG system must return valid JSON with `answer`, `sources`, and `confidence` fields. Before judging answer quality with `Correctness()`:
```python
scorers=[
    JsonValidity(schema={"required": ["answer", "sources", "confidence"]}),  # Tier 1
    NumericBound(min_value=0.0, max_value=1.0, field="confidence"),          # Tier 1
    IsNotEmpty(),                                                             # Tier 1
    Correctness(),                                                            # Tier 2 — skip if Tier 1 fails
]
```

**Scenario 2: Customer-facing chatbot with compliance requirements**
A financial services chatbot must include a disclaimer and stay under length limits:
```python
scorers=[
    ContainsKeywords(keywords=["not financial advice", "consult a professional"]),  # Policy gate
    LengthBound(max_length=500, unit="words"),                                      # Hard constraint
    RegexMatch(pattern=r"^(?!.*\b(guarantee|promise|certain)\b)"),                   # Safety gate
    Safety(),                                                                        # LLM judge
    Correctness(),                                                                   # LLM judge
]
```

**Scenario 3: Agentic tool-calling with latency SLAs**
An agent pipeline must respond within 3 seconds and produce non-empty output:
```python
scorers=[
    LatencyThreshold(max_latency_seconds=3.0),  # Tier 1 — from trace
    IsNotEmpty(),                                 # Tier 1
    ToolCallCorrectness(),                        # Tier 2
]
```

### Architecture

Introduce `BuiltInCodeScorer(Scorer)` — a lightweight base that provides serialization and registration without the LLM-specific requirements of `Judge`:

```python
class BuiltInCodeScorer(Scorer):
    """Base class for built-in deterministic code scorers."""
    name: str
    required_columns: set[str] = set()
    
    @property
    @abstractmethod
    def criteria(self) -> str:
        """Human-readable description of what this scorer checks."""
    
    @property
    def kind(self) -> ScorerKind:
        return ScorerKind.BUILTIN
    
    def _make_feedback(self, *, value: bool, rationale: str) -> Feedback:
        return Feedback(
            name=self.name,
            value=value,
            rationale=rationale,
            source=AssessmentSource(source_type="CODE", source_id=f"mlflow.scorers.{self.__class__.__name__}"),
        )
```

### Usage

```python
import mlflow
from mlflow.genai.scorers import (
    # Tier 1: Deterministic (instant, free)
    JsonValidity,
    ContainsKeywords,
    LengthBound,
    # Tier 2: LLM-based (nuanced, slower)
    Correctness,
    Safety,
)

results = mlflow.genai.evaluate(
    data=eval_data,
    predict_fn=my_model,
    scorers=[
        JsonValidity(),
        ContainsKeywords(keywords=["disclaimer"]),
        LengthBound(min_length=50, max_length=2000),
        Correctness(),
        Safety(),
    ],
)
```

### Files to change (Phase 1)

| File | Change |
|---|---|
| **`mlflow/genai/scorers/builtin_code_scorers.py`** | **[NEW]** `BuiltInCodeScorer` base + 8 scorer implementations |
| `mlflow/genai/scorers/__init__.py` | Add lazy imports + `__all__` exports |
| **`tests/genai/scorers/test_builtin_code_scorers.py`** | **[NEW]** Unit tests for all scorers |
| `docs/source/llms/genai/evaluation/index.rst` | Documentation section for deterministic scorers |

### Design decisions

- **`BuiltInCodeScorer(Scorer)` not `BuiltInCodeScorer(Judge)`**: Code scorers don't use LLMs and shouldn't carry LLM baggage.
- **Separate file**: `builtin_scorers.py` is 3000+ lines of LLM judges. Mixing in deterministic scorers would reduce clarity.
- **Reuses existing `CODE` source type and `BUILTIN` scorer kind**: No schema migration needed.
- **No new dependencies**: Uses Python stdlib (`re`, `json`) only.
- **All scorers return `Feedback` with rationale**: Parity with LLM judges in the UI.

## What machine learning domain(s) is this feature request about?

- [x] `domain/genai`: LLMs, Agents, and other GenAI-related use cases

## What area(s) of MLflow is this feature request about?

- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

### PR DESCRIPTION
**Pull request by dgenio**
### Related Issues/PRs

Closes #20827

### What changes are proposed in this pull request?

This PR implements Phase 1 of issue #20827: 9 built-in deterministic `BuiltInCodeScorer` implementations that complement the existing LLM judges with instant, free, and reproducible structural and safety checks.

#### New file: `mlflow/genai/scorers/builtin_code_scorers.py`

Introduces `BuiltInCodeScorer(Scorer)` — a lightweight base class that extends `Scorer` directly (not `Judge`) since these scorers require no LLM. The base provides serialization/deserialization via `SerializedScorer`, returns `ScorerKind.BUILTIN`, and implements `validate_columns()` for parity with `BuiltInScorer`.

The 9 new scorers:

| Scorer | What it checks | Key parameters |
|---|---|---|
| `ExactMatch` | Output equals `expected_response` | — |
| `JsonValidity` | Output is valid JSON, optionally with required top-level keys | `required_keys` |
| `RegexMatch` | Output matches a regex pattern | `pattern`, `full_match` |
| `ContainsKeywords` | All keywords present in output | `keywords`, `case_sensitive` |
| `LengthBound` | Output length within min/max | `min_length`, `max_length`, `unit` (`chars`/`words`) |
| `IsNotEmpty` | Output is non-empty and non-whitespace | — |
| `LatencyThreshold` | Trace latency ≤ threshold | `max_latency_seconds` |
| `NumericBound` | Numeric output within range (rejects NaN/inf) | `min_value`, `max_value` |
| `PII` | No PII detected (returns `True` = pass when clean) | `pii_types` (`email`, `phone`, `ssn`, `credit_card`) |

All scorers return `Feedback` with `AssessmentSourceType.CODE` and an informative rationale string, giving UI parity with LLM judges.

**Notable implementation details:**
- `NumericBound` uses `math.isfinite()` to explicitly reject NaN, inf, and -inf values.
- `PII` credit card detection uses Luhn checksum validation after regex matching to eliminate false positives from arbitrary digit sequences.

#### Changes to `mlflow/genai/scorers/__init__.py`

- Added `_LAZY_CODE_SCORER_IMPORTS` set mirroring the existing lazy-import pattern
- Extended `__getattr__`, `__dir__`, `TYPE_CHECKING` block, and `__all__` to export all 9 new scorers

#### Changes to `mlflow/genai/scorers/base.py`

- Added fallback in `Scorer.model_validate` to try `BuiltInCodeScorer.model_validate` when `BuiltInScorer.model_validate` raises for an unknown class name, so code scorers share the same deserialization entry point
- The fallback is narrowly scoped: only "Unknown builtin scorer class" errors fall through; other `MlflowException` errors (corrupted data, invalid structure) propagate correctly

#### New file: `tests/genai/scorers/test_builtin_code_scorers.py`

72 unit tests covering all 9 scorers: happy path, `None`/empty inputs, boundary values, error paths, non-finite numeric handling, credit card Luhn validation, column validation, and 3 serialization round-trips.

#### Usage example

```python
from mlflow.genai.scorers import (
    JsonValidity, ContainsKeywords, LengthBound, IsNotEmpty,
    Correctness, Safety,
)

results = mlflow.genai.evaluate(
    data=eval_data,
    predict_fn=my_model,
    scorers=[
        IsNotEmpty(),
        JsonValidity(required_keys=["answer", "sources"]),
        LengthBound(max_length=500, unit="words"),
        ContainsKeywords(keywords=["not financial advice"]),
        Correctness(),
        Safety(),
    ],
)
```

**Out of scope (Phase 2):** Gating infrastructure — running code scorers before LLM judges and skipping expensive judges on structural failures. Tracked in #20827.

### How is this PR tested?

- [x] New unit/integration tests

```
uv run --frozen pytest tests/genai/scorers/test_builtin_code_scorers.py -v
# 72 passed in 16.84s

uv run --frozen pytest tests/genai/scorers/test_serialization.py -v
# 20 passed — no regressions

uv run --frozen ruff check (all 4 changed files)
# All checks passed

uv run --frozen ruff format --check (all 4 changed files)
# All files already formatted
```

### Does this PR require documentation update?

- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

The 9 new scorer classes follow the same docstring style as existing built-in scorers and will appear in autodoc via `__all__`. A dedicated docs page (like the existing per-scorer pages under `docs/docs/genai/eval-monitor/scorers/`) can be added as a follow-up once the API is reviewed.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add 9 built-in deterministic code scorers (`ExactMatch`, `JsonValidity`, `RegexMatch`, `ContainsKeywords`, `LengthBound`, `IsNotEmpty`, `LatencyThreshold`, `NumericBound`, `PII`) to `mlflow.genai.scorers`. These scorers run without LLM calls — instant, free, and reproducible — and can be mixed with existing LLM judges in `mlflow.genai.evaluate()`.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

File: mlflow/genai/scorers/builtin_code_scorers.py [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+                value=False,
+                rationale=f"Numeric value {numeric_value} is " + " and ".join(violations) + ".",
+            )
+        return self._make_feedback(
+            value=True,
+            rationale=f"Numeric value {numeric_value} is within bounds"
+            + (f" (min={self.min_value})" if self.min_value is not None else "")
+            + (f" (max={self.max_value})" if self.max_value is not None else "")
+            + ".",
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply mlflow/genai/scorers/builtin_code_scorers.py:508]: Good catch! `NaN` comparisons silently return `False` in Python, so `NumericBound` would let NaN, inf, and -inf slip through bounds checks undetected.

Fixed in f908b6926 — added a `math.isfinite()` guard right after `float()` conversion. Non-finite values now return `value=False` with rationale `"Output ... is not a finite number."`. Also added tests for all three cases (NaN, inf, -inf).

File: mlflow/genai/scorers/builtin_code_scorers.py [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+        *,
+        outputs: Any = None,
+        expectations: dict[str, Any] | None = None,
+    ) -> Feedback:
+        output_str = _to_str(outputs)
+        expected = ""
+        if isinstance(expectations, dict):
+            expected = _to_str(expectations.get("expected_response", ""))
+        else:
+            expected = _to_str(expectations)
```
[the remark this answers is not in the document]
Comment (dgenio) [reply mlflow/genai/scorers/builtin_code_scorers.py:163]: Fair observation! The annotation says `dict[str, Any] | None` but the code also handles raw strings via the `else` branch. I'm keeping it as-is for now because `dict[str, Any] | None` is the convention used by other scorers in this package (e.g., in builtin_scorers.py). The string fallback is intentional tolerance for convenience, not a documented API path — so the annotation correctly describes the *expected* input contract while the code is simply permissive about what it accepts.
